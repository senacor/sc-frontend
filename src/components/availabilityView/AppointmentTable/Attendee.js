import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles/index';
import moment from 'moment-timezone';
import {
  appointmentsFilter,
  transformAppointmentTimeToPercent,
  styles
} from './AppointmentUtilities';

export function createSingleAppointmentDiv(
  distanceFromLeft,
  startTime,
  endTime,
  date,
  className
) {
  let timeStart = moment(startTime, 'YYYY-MM-DDTHH:mmZ[UTC]')
    .tz('Europe/Berlin')
    .format('HH:mm');
  let timeEnd = moment(endTime, 'YYYY-MM-DDTHH:mmZ[UTC]')
    .tz('Europe/Berlin')
    .format('HH:mm');
  let topPosition = transformAppointmentTimeToPercent(startTime, date);
  let length = transformAppointmentTimeToPercent(endTime, date) - topPosition;
  return (
    <div
      id={`appointmentDiv${distanceFromLeft}_${timeStart}-${timeEnd}`}
      key={`appointmentDiv${distanceFromLeft}_${timeStart}-${timeEnd}`} //needs an unique key
      className={className}
      style={{
        left: distanceFromLeft.toString() + '%',
        top: topPosition.toString() + '%',
        height: length.toString() + '%'
      }}
    />
  );
}

class Attendee extends React.Component {
  createAppointmentDivs() {
    let appointmentDivs = [];
    let filteredAppointments = appointmentsFilter(
      this.props.appointments,
      this.props.selectedDate
    );
    for (let i = 0; i < filteredAppointments.length; i++) {
      const startTime = filteredAppointments[i][0];
      const endTime = filteredAppointments[i][1];
      appointmentDivs.push(
        createSingleAppointmentDiv(
          this.props.distanceFromLeft,
          startTime,
          endTime,
          this.props.selectedDate,
          this.props.classes.appointmentDiv
        )
      );
    }
    return appointmentDivs;
  }

  render() {
    return (
      <React.Fragment>
        {this.props.show ? this.createAppointmentDivs() : null}
      </React.Fragment>
    );
  }
}

Attendee.propTypes = {
  appointments: PropTypes.array.isRequired,
  classes: PropTypes.object.isRequired,
  distanceFromLeft: PropTypes.number.isRequired,
  show: PropTypes.bool.isRequired
};

Attendee.defaultProps = {
  show: true
};

export const StyledComponent = withStyles(styles)(Attendee);
export default StyledComponent;
