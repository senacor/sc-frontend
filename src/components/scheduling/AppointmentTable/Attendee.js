import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment-timezone';
import { injectIntl } from 'react-intl';
import { withStyles } from '@material-ui/core/styles';
import Tooltip from '@material-ui/core/Tooltip';

import {
  appointmentsFilter,
  transformAppointmentTimeToPercent
} from './AppointmentUtilities';

const styles = theme => ({
  appointmentDiv: {
    width: '18%',
    borderRadius: 3,
    background: theme.palette.primary['300'],
    position: 'absolute'
  },
  appointmentDivEmployee: {
    width: '18%',
    borderRadius: 3,
    background: theme.palette.primary['400'],
    position: 'absolute'
  },
  appointmentDivReviewer: {
    width: '18%',
    borderRadius: 3,
    background: theme.palette.primary['500'],
    position: 'absolute'
  },
  appointmentDivSupervisor: {
    width: '18%',
    borderRadius: 3,
    background: theme.palette.secondary.green,
    position: 'absolute'
  }
});

export const createSingleAppointmentDiv = (
  distanceFromLeft,
  startTime,
  endTime,
  date,
  className,
  keySuffix,
  name,
  appointmentState,
  intl
) => {
  let timeStart = moment(startTime, 'YYYY-MM-DDTHH:mmZ[UTC]')
    .tz('Europe/Berlin')
    .format('HH:mm');
  let timeEnd = moment(endTime, 'YYYY-MM-DDTHH:mmZ[UTC]')
    .tz('Europe/Berlin')
    .format('HH:mm');
  let topPosition = transformAppointmentTimeToPercent(startTime, date);
  let length = transformAppointmentTimeToPercent(endTime, date) - topPosition;
  let infoOnMouseOver = name;
  switch (appointmentState) {
    case 'Busy':
      infoOnMouseOver = `${name}: ${intl.formatMessage({
        id: 'attendee.busy'
      })}`;
      break;
    case 'Tentative':
      infoOnMouseOver = `${name}: ${intl.formatMessage({
        id: 'attendee.tentative'
      })}`;
      break;
    case 'OOF':
      infoOnMouseOver = `${name}: ${intl.formatMessage({
        id: 'attendee.absent'
      })}`;
      break;
    case 'Free':
      infoOnMouseOver = `${name}: ${intl.formatMessage({
        id: 'attendee.free'
      })}`;
      break;
    default:
  }

  return (
    <Tooltip
      title={infoOnMouseOver}
      key={`appointmentDiv${distanceFromLeft}_${timeStart}-${timeEnd}-${keySuffix}`}
    >
      <div
        id={`appointmentDiv${distanceFromLeft}_${timeStart}-${timeEnd}`}
        key={`appointmentDiv${distanceFromLeft}_${timeStart}-${timeEnd}-${keySuffix}`} //needs an unique key
        className={className}
        style={{
          left: distanceFromLeft.toString() + '%',
          top: topPosition.toString() + '%',
          height: length.toString() + '%'
        }}
      />
    </Tooltip>
  );
};

const attendeeAppointmentClass = (attendee, classes) => {
  switch (attendee) {
    case 'employee':
      return classes.appointmentDivEmployee;
    case 'reviewer':
      return classes.appointmentDivReviewer;
    case 'supervisor':
      return classes.appointmentDivSupervisor;
    default:
      return classes.appointmentDiv;
  }
};

const Attendee = ({
  appointments,
  selectedDate,
  attendee,
  classes,
  distanceFromLeft,
  name,
  intl
}) => {
  const createAppointmentDivs = () => {
    let appointmentDivs = [];
    let filteredAppointments = appointmentsFilter(appointments, selectedDate);
    for (
      let appointmentCounter = 0;
      appointmentCounter < filteredAppointments.length;
      appointmentCounter++
    ) {
      const startTime = filteredAppointments[appointmentCounter][0];
      const endTime = filteredAppointments[appointmentCounter][1];
      const appointmentState = filteredAppointments[appointmentCounter][2];
      const className = attendeeAppointmentClass(attendee, classes);
      appointmentDivs.push(
        createSingleAppointmentDiv(
          distanceFromLeft,
          startTime,
          endTime,
          selectedDate,
          className,
          appointmentCounter,
          name,
          appointmentState,
          intl
        )
      );
    }
    return appointmentDivs;
  };

  return <React.Fragment>{createAppointmentDivs()}</React.Fragment>;
};

Attendee.propTypes = {
  appointments: PropTypes.array.isRequired,
  classes: PropTypes.object.isRequired,
  distanceFromLeft: PropTypes.number.isRequired,
  show: PropTypes.bool.isRequired,
  attendee: PropTypes.string
};

Attendee.defaultProps = {
  show: true
};

export default injectIntl(withStyles(styles)(Attendee));
