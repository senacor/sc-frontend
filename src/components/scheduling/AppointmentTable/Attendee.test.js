import React from 'react';
import { shallow } from 'enzyme';
import { StyledComponent as Attendee } from './Attendee';
import { createSingleAppointmentDiv } from './Attendee';
import moment from 'moment-timezone';
import Tooltip from '@material-ui/core/Tooltip';

const germanTimeFrom = dateTimeString =>
  moment
    .utc(dateTimeString, 'YYYY-MM-DDTHH:mmZ[UTC]')
    .tz('Europe/Berlin')
    .format('HH:mm');

const classes = {
  appointmentDiv: {
    color: '#AABBCC'
  }
};

const properties = {
  selectedDate: '2018-03-08',
  distanceFromLeft: 15.5,
  show: false,
  classes
};

describe('createSingleAppointmentDiv', () => {
  it('should create an appointmentDiv with the correct height and position.', () => {
    let startAppointment = '2018-03-08T09:45Z[UTC]';
    let endAppointment = '2018-03-08T15:15Z[UTC]';
    let appointment = [];
    appointment.push(startAppointment);
    appointment.push(endAppointment);
    let name = 'Michaela Mitarbeiterin';
    let appointmentState = 'Busy';
    let singleAppointmentDiv = (
      <Tooltip
        key={`appointmentDiv15.5_${germanTimeFrom(
          startAppointment
        )}-${germanTimeFrom(endAppointment)}-1`}
        title={name + ': beschÃ¤ftigt'}
      >
        <div
          key={`appointmentDiv15.5_${germanTimeFrom(
            startAppointment
          )}-${germanTimeFrom(endAppointment)}-1`} //not visible, but necessary for equality
          className={classes.appointmentDiv}
          id={`appointmentDiv15.5_${germanTimeFrom(
            startAppointment
          )}-${germanTimeFrom(endAppointment)}`}
          style={{
            left: '15.5%',
            top: '25%',
            height: '50%'
          }}
        />
      </Tooltip>
    );
    expect(
      createSingleAppointmentDiv(
        15.5,
        startAppointment,
        endAppointment,
        properties.selectedDate,
        classes.appointmentDiv,
        1,
        name,
        appointmentState
      )
    ).toEqual(singleAppointmentDiv);
  });
});

describe('createAppointmentDivs', () => {
  it('should create two appointmentDivs for two appointments.', () => {
    let startAppointment = '2018-03-08T09:45Z[UTC]';
    let endAppointment = '2018-03-08T15:15Z[UTC]';
    let appointmentStatus = 'Tentative';
    let appointment = [];
    appointment.push(startAppointment);
    appointment.push(endAppointment);
    appointment.push(appointmentStatus);
    let appointmentsEmployee = [];
    appointmentsEmployee.push(appointment);
    appointmentsEmployee.push(appointment);
    const component = shallow(
      <Attendee
        appointments={appointmentsEmployee}
        selectedDate={properties.selectedDate}
        distanceFromLeft={properties.distanceFromLeft}
        show={false}
      />
    ).dive();
    component.setProps({ classes });

    expect(component.instance().createAppointmentDivs()).toHaveLength(2);
  });
});

describe('Attendee', () => {
  it('renders appointments if show is set to true.', () => {
    let startAppointment = '2018-03-08T09:45Z[UTC]';
    let endAppointment = '2018-03-08T15:15Z[UTC]';
    let appointmentStatus = 'Tentative';
    let appointment = [];
    appointment.push(startAppointment);
    appointment.push(endAppointment);
    appointment.push(appointmentStatus);
    let appointmentsEmployee = [];
    appointmentsEmployee.push(appointment);
    appointmentsEmployee.push(appointment);
    let name = 'Michaela Mitarbeiterin';
    const component = shallow(
      <Attendee
        appointments={appointmentsEmployee}
        selectedDate={properties.selectedDate}
        distanceFromLeft={properties.distanceFromLeft}
        show={true}
      />
    ).dive();
    component.setProps({ classes, name });

    expect(component).toMatchSnapshot();
  });
});
