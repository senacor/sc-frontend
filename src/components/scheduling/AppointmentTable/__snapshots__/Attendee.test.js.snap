// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attendee renders appointments if show is set to true. 1`] = `
<Fragment>
  <div
    id="appointmentDiv15.5_10:45-16:15"
    key="appointmentDiv15.5_10:45-16:15-0"
    style={
      Object {
        "height": "50%",
        "left": "15.5%",
        "top": "25%",
      }
    }
  />
  <div
    id="appointmentDiv15.5_10:45-16:15"
    key="appointmentDiv15.5_10:45-16:15-1"
    style={
      Object {
        "height": "50%",
        "left": "15.5%",
        "top": "25%",
      }
    }
  />
</Fragment>
`;
